tinygrad/shape/view.py

```diff
@@ -15,6 +15,58 @@ def strides_for_shape(shape:Tuple[int, ...]) -> Tuple[int, ...]:
   for d in reversed(shape[1:]): strides.append(d*strides[-1])
   return filter_strides(shape, tuple(reversed(strides)))
 
+@functools.lru_cache(maxsize=None)
+def to_shape_strides(shape:Tuple[int, ...], strides:Tuple[int, ...], mask:Optional[Tuple[Tuple[int, int], ...]] = None) -> Tuple[Tuple[int, int, int], ...]:
+  assert len(shape) == len(strides)
+  state = 1 if mask and strides[0] == 0 and mask[0][1] - mask[0][0] == 1 and shape[0] != 1 else 0
+  ret = [(shape[0], strides[0], shape[0] if strides[0] else 0)] if shape else []
+  for i in range(1, len(shape)):
+    if shape[i] == 1:
+      continue
+    if mask and strides[i] == 0 and mask[i][1] - mask[i][0] == 1:
+      if state == 1:
+        ret[-1] = (ret[-1][0] * shape[i], 0, 0)
+      else:
+        ret.append((shape[i], 0, 0))
+      state = 1
+    else:
+      if state == 1 or ret[-1][1] == shape[i] * strides[i]:
+        ret[-1] = (ret[-1][0] * shape[i], strides[i], (shape[i] if state == 1 else ret[-1][2] * shape[i]) if strides[i] else 0)
+      else:
+        ret.append((shape[i], strides[i], shape[i] if strides[i] else 0))
+      state = (2 if state != 0 else 0) 
+      
+  return tuple(ret)
+
+@functools.lru_cache(maxsize=None)
+def _reshape_mask(view: View, new_shape:Tuple[sint, ...]) -> Tuple[Optional[Tuple[Tuple[sint, sint], ...]], Optional[Tuple[sint, ...]], bool]:
+  if view.mask is None: return view.mask, tuple(), False
+  new_mask: List[Tuple[int, int]] = []
+  r_masks, r_shape, r_new_shape = reversed(view.mask), reversed(view.shape), reversed(new_shape)
+  # stride represents the curr stride while accumulating. offset is needed while combining masks of range one & zero stride
+  # View(shape=(10, 10), strides=(0, 1), offset=0, mask=((3, 4), (1, 5)), contiguous=False).reshape((100,))
+  stride, off, offsets, old_dim, new_dim, mask = 1, 0, [], next(r_shape, 1), next(r_new_shape, 1), next(r_masks, (0,1))
```

Comment:
is there a better way to write the end condition for these iterator (like a None or catch StopIteration)?